### PR TITLE
 DAT-36: BE - Get list user 

### PR DESCRIPTION
**Pull request by thanhphuchuynh**
**Add get list user  API** 
Required token in the header
Structure of the result: 
```
{
    "totalItems": 10,
    "totalPages": 1,
    "currentPage": 1,
    "maxItemsPerPage": 10,
    "listUsers": []
}
```
- `/users`:  get all users, because it is a list of all users, so totalItem = numbers users and totalPage = 1
image

- `users?page=2&size=4` :  get list users by page, size, 4 user per one page
image

-` /users?size=4`: default page = 1
image

### CODE REVIEW

File: internal/app/api/services/user/user.go [hunk 1]
```diff
@@ -145,3 +147,48 @@ func (s *Service) UpdateUserByID(ctx context.Context, user types.User) error {
...
+
+// Get list users by page
+func (s *Service) GetListUsers(ctx context.Context, page, size string) (*types.GetListUsersResponse, error) {
+
+	var pagingNSorting types.PagingNSorting
+	err := pagingNSorting.Init(page, size)
+
+	if err != nil {
+		s.logger.Errorf("Failed url parameters when get list users", err)
+		return nil, err
```
Comment (paphuc) [inline internal/app/api/services/user/user.go:159]: Create a new custom error message then return, the err from the repo is just to log

File: internal/app/api/services/user/user.go [hunk 2]
```diff
@@ -145,3 +147,48 @@ func (s *Service) UpdateUserByID(ctx context.Context, user types.User) error {
...
+
+	if pagingNSorting.Size > numberUsers {
+		listUsersResponse.MaxItemsPerPage = numberUsers
+	}
+
+	listUsers, err := s.repo.GetListUsers(ctx, pagingNSorting)
+
+	if err != nil {
+		s.logger.Errorf("Failed when get list users by page", err)
+		return nil, err
```
Comment (paphuc) [inline internal/app/api/services/user/user.go]: Create a new custom error message to return

File: internal/app/api/services/user/user.go [hunk 3]
```diff
@@ -145,3 +147,48 @@ func (s *Service) UpdateUserByID(ctx context.Context, user types.User) error {
...
+	}
+
+	listUsersResponse.ListUsers = append(listUsersResponse.ListUsers, listUsers...)
+
+	s.logger.Infof("get list users by page is completed, page: ", pagingNSorting)
+	return &listUsersResponse, err
```
Comment (paphuc) [inline internal/app/api/services/user/user.go]: should return `&listUsersResponse, nil` for better understanding

File: internal/app/api/types/pagesort.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import "strconv"
+
+type PagingNSorting struct {
+	Size int `json:"size" default:"100"`
+	Page int `json:"page" default:"1"`
+}
+
+func (ps *PagingNSorting) Init(page, size string) error {
+
+	if size == "" && page == "" {
```
Comment (paphuc) [inline internal/app/api/types/pagesort.go:19]: Have you tested this case?

File: internal/app/api/types/pagesort.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+		ps.Size = 100
+		ps.Page = 1
+		return nil
+	}
+
+	sizeInt, err := strconv.ParseInt(size, 10, 64)
```
Comment (paphuc) [inline internal/app/api/types/pagesort.go:25]: What if size has a value, but page not?

File: internal/app/api/types/user.go [hunk 6]
```diff
@@ -59,7 +60,14 @@ type UserResponseSignUp struct {
 	Email string `json:"email"`
 	Token string `json:"token"`
 }
+type GetListUsersResponse struct {
+	TotalItems      int               `json:"totalItems"`
+	TotalPages      int               `json:"totalPages"`
+	CurrentPage     int               `json:"currentPage"`
```
Comment (paphuc) [inline internal/app/api/types/user.go]: Please create a struct for these fields, then replace them

File: internal/app/api/types/pagesort.go (hunk 4 above)
Comment (thanhphuchuynh) [reply internal/app/api/types/pagesort.go:19]: this gets all user
image
or
image

File: internal/app/api/types/pagesort.go (hunk 5 above)
Comment (thanhphuchuynh) [reply internal/app/api/types/pagesort.go:25]: the default value of page is 1 (logic in line 21)
image

**Pull request merged**